### PR TITLE
Add clang-format.

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

**Comment by dalg24** [comment]
I need to make some changes to `diff-clang-format.py` in Cap to be able to specify what version of clang-format to use.

**Comment by Rombur** [comment]
@dalg24 @aprokop @sslattery  please tell me what you think of this format

**Comment by aprokop** [comment]
@Rombur I'll try to this by next week.

File: cmake/CodeFormat.cmake [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            --config=${${PACKAGE_NAME}_SOURCE_DIR}/.clang-format
+            --apply-patch
+            ${${PACKAGE_NAME}_SOURCE_DIR}/packages
+    )
+    file(WRITE
+        ${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/check_format_cpp.sh
+        "#!/usr/bin/env bash\n"
+        "\n"
+        "${PYTHON_EXECUTABLE} "
+        "${CMAKE_BINARY_DIR}/diff-clang-format.py "
```
Comment (aprokop) [inline cmake/CodeFormat.cmake]: diff-clang-format.py seems to produce counter-intuitive diff. It seems to first print the new version with <, and then the old version with >. This is not how the regular diff works. Instead, it should first print old version with <, and then the new version with >.

Comment (aprokop) [inline cmake/CodeFormat.cmake]: Another cool feature would be to have something like "diff -u" += style.

Comment (dalg24) [inline cmake/CodeFormat.cmake]: c.f. ORNL-CEES/Cap#234

**Comment by sslattery** [comment]
Is it possible to have no spaces between commas in template parameters?

i.e. SomeClass<double,int,long> instead of SomeClass<double, int, long>

**Comment by dalg24** [comment]
Looking at the [doc], I do not see how you would specify that.

**Comment by Rombur** [comment]
@sslattery unfortunately no. There needs to be a space after a comma.

**Comment by sslattery** [comment]
OK - no problem. That one is not a big deal

**Comment by aprokop** [comment]
If this is something we really want, we could add our own rule for that. Something like: "if < is the first non-whitespace in a row, and > is the last non-whitespace in a row, and the row has no other special symbols besides commas, get rid of spaces".

**Comment by dalg24** [comment]
@aprokop I'm not sure you can define your custom rule with clang-format.

**Comment by sslattery** [comment]
This one is not a problem. Its one of the few differences I've noted so far so just thought I would check

**Comment by aprokop** [comment]
@dalg24 I did not mean a custom rule with clang-format (though in the long term it would be nice), but rather run clang-format first, and then post-process its output.

**Comment by dalg24** [comment]
Yuk!

**Comment by sslattery** [comment]
I don't like how its doing function arguments and initializer lists. We should talk about this one. DTK_ParallelSearch.cpp is a good example

**Comment by dalg24** [comment]
@Rombur @sslattery  I rebased this on master.

**Comment by Rombur** [comment]
@sslattery I have changed the indentation for the constructor and now it is the same as the current master (one element per line and it start with the comma). But I can't easily change it for the other functions. If it really bothers you I can look into it more but it will require to define a break condition for every parenthesis, which I don't want to do.

**Comment by Rombur** [comment]
OK this is ready to be merged

**Comment by dalg24** [comment]
HOLD ON

**Comment by dalg24** [comment]
I'm not happy with squashing everything in 3d6bd20.  Please keep the actual `make format-cpp` in a separate commit.

**Comment by dalg24** [comment]
@Rombur I didn't pull so if you can either cherry pick or I can share the previous version with you and just update `.clang-format` and run `make-format`again.

**Comment by dalg24** [comment]
OK.  I fixed it.

**Comment by dalg24** [comment]
This closes #77

**Pull request merged**